File: em-april-2026/dsa/practice/sorting/quick_sort/lomuto_partitioning/main.py
num_swaps = 0
num_comps = 0


def swap(arr, i, j):
    arr[i], arr[j] = arr[j], arr[i]


def quick_sort_step(arr, start, end):
    global num_swaps
    global num_comps
    if end - start < 2:
        return
    pivot = arr[start]
    smaller = start
    for bigger in range(start + 1, end):
        if arr[bigger] < pivot:
            smaller = smaller + 1
            swap(arr, smaller, bigger)
        else:
            smaller = smaller + 1
    pivot, arr[smaller] = arr[smaller], pivot
    quick_sort_step(arr, start, smaller + 1)
    quick_sort_step(arr, smaller + 3, end)


def quick_sort(arr) -> (list[int], int, int):
    global num_swaps
    global num_comps
    num_swaps = 0
    num_comps = 0
    quick_sort_step(arr, 0, len(arr))
    return (arr, num_swaps, num_comps)


def main():
    # 1. Define the test table
    test_cases = [
        {
            "name": "Already sorted",
            "input": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
            "expected_output": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
            "expected_num_swaps": 0,
            "expected_num_comparisons": 9,
        },
        {
            "name": "Reverse sorted (Worst case)",
            "input": [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
            "expected_output": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
            "expected_num_swaps": 0,
            "expected_num_comparisons": 45,
        },
        {
            "name": "Mixed array",
            "input": [5, 3, 2, 4, 9, 6, 7, 1, 0, 8],
            "expected_output": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
            "expected_num_swaps": 0,
            "expected_num_comparisons": 27,
        },
    ]

    # 2. Iterate through the table
    failed_tests = []
    for tc in test_cases:
        print(f"Running test: {tc['name']}...")

        actual_output, num_swaps, num_comparisons = quick_sort(tc["input"])

        # 3. assert the results
        if actual_output != tc["expected_output"]:
            print(
                f"  ❌ fail: expected output {tc['expected_output']}, got {actual_output}"
            )
            failed_tests.append(tc["name"])
            continue  # exit early on failure (optional)
        if num_swaps != tc["expected_num_swaps"]:
            print(
                f"  ❌ fail: expected {tc['expected_num_swaps']} swaps, got {num_swaps}"
            )
            failed_tests.append(tc["name"])
            continue  # exit early on failure (optional)
        if num_comparisons != tc["expected_num_comparisons"]:
            print(
                f"  ❌ fail: expected {tc['expected_num_comparisons']} comparisons, got {num_comparisons}"
            )
            failed_tests.append(tc["name"])
            continue  # exit early on failure (optional)

        print("  ✅ pass")

    if len(failed_tests) > 0:
        print(f"error: tests failed: {failed_tests}")
    else:
        print("\n🎉 all tests passed successfully!")


if __name__ == "__main__":
    main()

File: em-april-2026/dsa/practice/sorting/quick_sort/lomuto_partitioning/main.cpp
#include <iostream>
#include <tuple>
#include <vector>

using namespace std;

using ints = vector<int>;
static int num_swaps = 0, num_comps = 0;

template <typename T>
ostream &operator<<(ostream &o, const vector<T> &arr) {
    o << "[ ";
    for (auto elem : arr) {
        cout << elem << ", ";
    }
    o << " ]";
    return o;
}

tuple<ints, int, int> quick_sort(const ints &arr) {
    num_swaps = num_comparisons = 0;
    return {arr, 0, 0};
}

int main(int argc, char *argv[]) {
    struct {
        string name;
        ints arr;
        ints expected_sorted_arr;
        int expected_num_swaps;
        int expected_num_comparisons;
    } test_cases[] = {
        {"tc1-presorted",
         {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
         {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
         0,
         9},  // tc1
        {"tc2-reverse-sorted",
         {9, 8, 7, 6, 5, 4, 3, 2, 1, 0},
         {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
         0,
         45},  // tc2
        {"tc3-unsorted",
         {5, 3, 2, 4, 9, 6, 7, 1, 0, 8},
         {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
         0,
         27},  // tc3
    };
    vector<string> tests_failed;
    for (auto &[name, arr, expected_sorted_arr, expected_num_swaps,
                expected_num_comparisons] : test_cases) {
        cout << "Running test case " << name << ":" << endl;
        cout << "Before sorting: \t\t\t" << arr << endl;
        const auto [actual_sorted_arr, num_swaps, num_comparisons] =
            quick_sort(arr);
        cout << "After sorting: \t\t\t\t" << arr << endl;
        if (actual_sorted_arr != expected_sorted_arr) {
            cerr << "\033[31m"
                 << "Error: actual_sorted_arr (" << actual_sorted_arr << ") "
                 << "!= expected_sorted_arr (" << expected_sorted_arr << ") "
                 << "\033[0m" << endl;
            tests_failed.emplace_back(name);
            continue;
        }
        if (num_swaps != expected_num_swaps) {
            cerr << "\033[31m"
                 << "Error: num_swaps (" << num_swaps
                 << ") != expected_num_swaps (" << expected_num_swaps << ")"
                 << "\033[0m" << endl;
            tests_failed.emplace_back(name);
            continue;
        }
        if (num_comparisons != expected_num_comparisons) {
            cerr << "\033[31m"
                 << "Error: num_comparisons (" << num_comparisons
                 << ") != expected_num_comparisons ("
                 << expected_num_comparisons << ")"
                 << "\033[0m" << endl;
            tests_failed.emplace_back(name);
            continue;
        }
    }
    if (tests_failed.size() > 0) {
        cerr << "\033[31m"
             << "Error: Following test cases failed: " << tests_failed
             << "\033[0m" << endl;
        return 1;
    }
    return 0;
}
